# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [78]:
import pandas as pd

In [79]:
df = pd.read_csv('../data/supermarket_sales.csv')  
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [80]:
df.sort_values(by = 'Unit price', ascending = False).head() 


Invoice ID Branch       City Customer type Gender         Product line  \
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
122  899.64                 4.761905        44.982     4.2  
983  699.72                 4.761905        34.986     6.1  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [81]:
product_el = df[df['Product line'] == 'Electronic accessories']  
max_quantity = product_el.loc[product_el['Quantity'].idxmax()]  
display(max_quantity)

Invoice ID                            732-94-0499
Branch                                          C
City                                    Naypyitaw
Customer type                              Normal
Gender                                     Female
Product line               Electronic accessories
Unit price                                  41.65
Quantity                                       10
Tax 5%                                     20.825
Total                                     437.325
Date                                    1/13/2019
Time                                        17:04
Payment                               Credit card
cogs                                        416.5
gross margin percentage                  4.761905
gross income                               20.825
Rating                                        5.4
Name: 75, dtype: object

In [82]:
max_price = product_el.loc[product_el['Unit price'].idxmax()]  
display(max_price)

Invoice ID                            817-69-8206
Branch                                          B
City                                     Mandalay
Customer type                              Normal
Gender                                     Female
Product line               Electronic accessories
Unit price                                  99.73
Quantity                                        9
Tax 5%                                    44.8785
Total                                    942.4485
Date                                     3/2/2019
Time                                        19:42
Payment                               Credit card
cogs                                       897.57
gross margin percentage                  4.761905
gross income                              44.8785
Rating                                        6.5
Name: 209, dtype: object

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [83]:
def rates(x):
    if x >= 8:
        return'High'
    elif 6 <= x < 8:
        return'Medium'
    else:
        return 'Low'

In [84]:
df['Rating Category'] = df.Rating.apply(rates)

In [85]:
df.head(10)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   
5  699-14-3026      C  Naypyitaw        Normal    Male   
6  355-53-5943      A     Yangon        Member  Female   
7  315-22-5665      C  Naypyitaw        Normal  Female   
8  665-32-9167      A     Yangon        Member  Female   
9  692-92-5582      B   Mandalay        Member  Female   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   
5  Electronic accessories       85.39         7  29.8865  627.6165  3/25/2019   
6  Electronic accessories       68.84         6  20.6520  433.6920  2/25/2019   
7      Home and lifestyle       73.56        10  36.7800  772.3800  2/24/2019   
8       Health and beauty       36.26         2   3.6260   76.1460  1/10/2019   
9      Food and beverages       54.84         3   8.2260  172.7460  2/20/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   
5  18:30      Ewallet  597.73                 4.761905       29.8865     4.1   
6  14:36      Ewallet  413.04                 4.761905       20.6520     5.8   
7  11:38      Ewallet  735.60                 4.761905       36.7800     8.0   
8  17:15  Credit card   72.52                 4.761905        3.6260     7.2   
9  13:27  Credit card  164.52                 4.761905        8.2260     5.9   

  Rating Category  
0            High  
1            High  
2          Medium  
3            High  
4             Low  
5             Low  
6             Low  
7            High  
8          Medium  
9             Low

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [86]:
rate_high = df[df['Rating Category'] == 'High'].count() / df['Rating Category'].count() * 100
print(rate_high['Rating Category'])

32.9


Категорія High становить 32.9%

In [87]:
rate_medium = df[df['Rating Category'] == 'Medium'].count() / df['Rating Category'].count() * 100 
print(rate_medium['Rating Category'])

35.6


Категорія Medium становить 35.6%

In [88]:
rate_low = df[df['Rating Category'] == 'Low'].count() / df['Rating Category'].count() * 100  
print(rate_low['Rating Category'])

31.5


Категорія Low становить 31.5%

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [89]:
time = '11:28'

In [93]:
def get_hour(x):
    time_split = x.split(':')
    return int(time_split[0])
    

In [94]:
get_hour(time)


11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [95]:
df['Hour'] = df.Time.apply(get_hour)  
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [96]:
df['Hour'].value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Найбільше покупок здійснюється о 19й годині, також доволі багато покупок здійснюється о 10й, 13й, та 15й годинах. Можно припустити,що о 19й годині люди повертаються з роботи та заходять і  купляють щось в магазині, це логічно. Щодо 13ї години - це час перерви,тому люди у час перерви заходять до магазину. За цією ж логікою, 15я година - це також час перерви або кінець робочого дня у деяких робітників, а 10а година - це початок робочого дня,тобто люди заходять до магазину перед початком роботи. 

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [98]:
df['Total'].groupby(by = df['City']).mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [100]:
df['Total'].groupby(by = df['City']).min()

City
Mandalay     18.6375
Naypyitaw    10.6785
Yangon       12.6945
Name: Total, dtype: float64

In [101]:
df['Total'].groupby(by = df['City']).max()

City
Mandalay     1022.49
Naypyitaw    1042.65
Yangon       1039.29
Name: Total, dtype: float64

In [102]:
df['Total'].groupby(by = df['City']).mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

In [112]:
df['Total'].groupby(by = df['City']).count()

City
Mandalay     332
Naypyitaw    328
Yangon       340
Name: Total, dtype: int64

Найбільша кількість покупок була здійснена у місті Yangon.При цьому місто Yangon має найменше середнє значення чеку в порівнянні з двома іншими містами.Місто Naypyitaw при найменшій кількості покупок має найвище середнє значення чеку, найвище максимальне значення чеку та найменше мінімальне значення чеку. Місто Mandalay має найвище мінімальне значення суми чеку, найменше максимальне значення та займає друге місце за середнім значенням чеку та кількості покупок серед усіх трьох міст. Можемо зробити висновок, що магазини у місті Naypyitaw працюють краще за інших,але можна попрацювати над залученням нових клієнтів та збільшенням кількості покупок у магазині (додаткові акції,клубні картки). Магазин у місті Yangon треба удосконалити - перевірити рейтинг, можливо підняти ціни на якусь категорію товарів (виявити яка категоря товарів має попит),провести маркетингові операції ,які піднімуть середнє значення суми чеку.У магазині міста Mandalay також можна попрацювати над підвищенням середьої суми чека - 

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [ ]:
df.to_csv('supermarket_sales_enriched.csv', index = False)